### PR TITLE
Vpn panel

### PR DESCRIPTION
**Pull request by ravjit-cliqz**

### CODE REVIEW

File: vpn/src/main/java/com/cliqz/browser/vpn/DisconnectVPN.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * Created by arne on 13.10.13.
+ */
+public class DisconnectVPN extends Activity implements DialogInterface.OnClickListener, DialogInterface.OnCancelListener {
+    private IOpenVPNServiceInternal mService;
+    private ServiceConnection mConnection = new ServiceConnection() {
+
+
+
```
Comment (jayeshsolanki93) [inline vpn/src/main/java/com/cliqz/browser/vpn/DisconnectVPN.java]: Random whitespace. There at a couple of places.

File: vpn/src/main/java/com/cliqz/browser/vpn/DisconnectVPN.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        unbindService(mConnection);
+    }
+
+    private void showDisconnectDialog() {
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
```
Comment (jayeshsolanki93) [inline vpn/src/main/java/com/cliqz/browser/vpn/DisconnectVPN.java]: final

File: vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+
+    public static void setExternalAuthProviderSpinnerList(Spinner spinner, String selectedApp) {
+        Context c = spinner.getContext();
+        final PackageManager pm = c.getPackageManager();
+        ArrayList<ExternalAuthProvider> extProviders = getExternalAuthProviderList(c);
+
+        int selectedPos = -1;
+
+        if (extProviders.size() ==0)
```
Comment (jayeshsolanki93) [inline vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java]: `isEmpty` instead?

File: vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            if (extProviders.get(i).packageName.equals(selectedApp))
+                selectedPos = i;
+        }
+        SpinnerAdapter extAppAdapter = new ArrayAdapter<ExternalAuthProvider>(c, android.R.layout.simple_spinner_item, android.R.id.text1, extProviders);
+        spinner.setAdapter(extAppAdapter);
+        if (selectedPos != -1)
+            spinner.setSelection(selectedPos);
+    }
+
+    static ArrayList<ExternalAuthProvider> getExternalAuthProviderList(Context c) {
```
Comment (jayeshsolanki93) [inline vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java]: Can return `List<ExternalAuthProvider>`. Programming to interface not implementation.

http://thefinestartist.com/effective-java/52

Comment (ravjit-cliqz) [reply vpn/src/main/java/com/cliqz/browser/vpn/core/ExtAuthHelper.java]: All the code inside vpn module is not mine.

**Pull request merged**